### PR TITLE
Address edge cases around multiple content types

### PR DESCRIPTION
**Pull request by matt9j**
## Description

This PR fixes to main issues in the theme with content types that are not "post":

Firstly, it adds these content types to the appropriate taxonomy pages for categories and tags, allowing them to be listed if they are tagged or categorized appropriately.

Secondly, it moves the content-list.html and its dependent templates to the _default layout type. This is important since the homepage is a list, and requires content to define a content-list template to render appropriately.

Lastly, it separates the rendering of the homepage from the rendering of list pages, to allow list pages to only render the pages in their section.

## Motivation and Context

This change fixes one of the underlying causes of bug #119 .

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have read the [Contributing Document](https://github.com/pacollins/hugo-future-imperfect-slim/blob/master/.github/CONTRIBUTING.md).
- [X] My code follows the [code style](https://github.com/pacollins/hugo-future-imperfect-slim/blob/master/.github/CONTRIBUTING.md#Style-Guide) of this project.
- [ ] My change requires a change to the [documentation](https://github.com/pacollins/hugo-future-imperfect-slim/wiki).
- [ ] I have updated the documentation, including `theme.toml`, accordingly.

### CODE REVIEW

**Comment by matt9j** [comment]
Addressing the issue of specific lists for each content type will require a larger change that I wasn't thinking would be in-scope for this PR. I will push a commit to this PR which adds a separate home template to achieve this behavior and completely address #119 if that is what is desired though. I think the more difficult question around that bug is not "how to do this," but rather "what is the intended behavior". If the intended behavior is to have all content types in `mainSections` appear on the homepage, but then still have separate subsections with only the section content type, then that can be achieved by adding a homepage layout that is distinct from the `layouts/_default/list.html` layout. I think this is the most "hugo-like" way to do things, rather than adding switching logic to list.html around the `.Page.isHome` parameter.

**Comment by pacollins** [comment]
I think the intent would be to have the homepage show all content types while each individual page only shows its content type. For added flexibility though, it would be ideal for the homepage to have the ability to select content types via config.

**Comment by matt9j** [comment]
> I think the intent would be to have the homepage show all content types while each individual page only shows its content type. For added flexibility though, it would be ideal for the homepage to have the ability to select content types via config.

Sounds good! That's now the behavior of this branch (I think!). mysiteurl/contenttype should display a list of pages with either explicit `type = "contenttype"` set in the front matter or inferred by being in the directory `content/contenttype/foo.md` The homepage should display all pages with a type (either inferred or explicit) that is in the `mainSections` config parameter list.

FYI there's been some more discussion in #119 also about how non-post single templates render, and I'm happy to incorporate those changes either here or in a followon pull request if needed too.

**Comment by pacollins** [comment]
I think that is applicable to this PR, so go ahead and add it here.

**Comment by matt9j** [comment]
> I think that is applicable to this PR, so go ahead and add it here.

Done!

**Comment by pacollins** [comment]
This might be an edge case that doesn't matter, but when the directory and type don't match it seems to be throwing the template off sometimes. Example:

Setting `content/test/item.md => type="item"` returns `http://localhost:1313/test/item1/ Recent content in Itemized Test on Hugo Future Imperfect Slim Hugo -- gohugo.io en Tue, 05 Mar 2019 00:00:00 +0000`

Interestingly enough, setting `content/fake/item.md => type="item"` works properly.

**Comment by matt9j** [comment]
Ah, interesting. I think I learned something about Hugo paths today and had an incorrect mental model of the behavior! I think there is also a bug with the hugo server development configuration. What follows is my current understanding and what I am able to consistently reproduce with version `v0.67.1/extended linux/amd64` (latest on Archlinux)

For context, this is my test environment content directory (based off my own site)
```
❯ tree content
content
├── about
│   └── _index.md
├── contact
│   └── _index.md
├── foo
│   └── test-foo.md
├── item
│   ├── test-item2.md
│   └── test-item.md
├── posts
│   └── hello-world.md
└── publications
    ├── chi2018.md
    ├── cscw2019.md
    ├── hotnets2018.md
    ├── _index.md
    ├── limits2017.md
    └── nsdi19.md
```
I have the theme set to matt9j:devel-multiple-content-types, and a custom layout defined for the type `publication` in mysite/layouts that has some subtle changes to how it's listed for long titles (but that I can tell apart from the default).  

Findings:
1) The url path layout is completely determined by the file layout of the content directory. Regardless of what `type` is set to in the front matter of `test-foo.md`, `test-item.md`, or `test-item2.md`, these posts will always be listed at mysiteurl/foo or mysiteurl/item appropriately. Explicitly setting `test-foo.md` to type `item` in the front matter still leads to it being listed at mysiteurl/foo and available as a single page at `mysiteurl/foo/test-foo/` This doc page explains this behavior, and the difference between a "section" and a "type" in hugo parlance. https://gohugo.io/content-management/sections/

2) The type will be used when resolving which layout template to use for the content. Setting the type of test-foo.md to `publication` (my custom layout) in the front matter will lead to it rendering with the publication specific content-list and single page templates as expected. The page will still be available in the list at `mysiteurl/foo/` and as a single page at `mysiteurl/foo/test-foo/` per finding 1. It will *not* be reachable at `mysiteurl/publication/test-foo` and will not appear in `mysiteurl/publication/`

3) The development server with change detect does not handle type changes well, even with the `--disableFastRender` option set :open_mouth: When using the debug server, if I change the frontmatter type of test-foo back and forth between foo, item, and publication, the debug server will detect the change and reload, but *will not render the list page at `mysiteurl/foo/` correctly* after that, no matter what the frontmatter type of test-foo is. Even removing the type specification from the front matter completely does not fix the list rendering for me. The entry always show up unrendered as a raw hugo path (i.e. `http://localhost:1313/foo/test-foo/ Recent content in Test Foo on matt9j:Matt Johnson Hugo -- gohugo.io en Mon, 16 Mar 2020 10:17:35 -0700`). *The only thing that fixes the list rendering is a complete restart of the hugo server*. After that it renders correctly, until I change the type again. To make things even stranger, the single page template at `mysiteurl/foo/test-foo` picks up the type changes correctly and correctly re-renders automatically!

TLDR:
I think this is a hugo issue, not a template issue. I am able to consistently get the correct behavior with a fresh run of hugo server, and reproduce the failure consistently as well. I will open a bug with hugo upstream if you're also able to reproduce this issue!

The first part of my previous assertion:
> Sounds good! That's now the behavior of this branch (I think!). mysiteurl/contenttype should display a list of pages with either explicit type = "contenttype" set in the front matter or inferred by being in the directory content/contenttype/foo.md The homepage should display all pages with a type (either inferred or explicit) that is in the mainSections config parameter list.
is incorrect though... with this PR the URL is determined by the location in the content directory, the rendering template is determined by the type, and homepage inclusion is determined by the page type as well.

**Comment by pacollins** [comment]
> It will not be reachable at `mysiteurl/publication/test-foo` and will not appear in `mysiteurl/publication/`

It should be reachable at `mysiteurl/publications/test-foo` and `mysiteurl/publications/` because your directory is `publications` not `publication`.

As for the TLDR, I suppose it does make more sense that the directory content is located under would not necessarily determine the template used. This would be unnecessarily limiting. What it _does_ mean is that we should probably remove `type=""` from the content where the directory name matches the template name because it is unnecessary when using the default template, correct? It would make sense that we would aim, as a theme development team, to develop in a way that pages reference the lowest point in the [look-up order](https://gohugo.io/templates/lookup-order/) so that users can add increased specificity. Right?

**Comment by matt9j** [comment]
> It should be reachable at mysiteurl/publications/test-foo and mysiteurl/publications/ because your directory is publications not publication

No actually. Sorry that directory tree is unnecessarily complex. I have all the publications in the publications directory so the URLs have correct english pluralization `mysite/publications/pubname` but they actually all have type `publication` set in their front matter, and I have layout templates defined in my site (not in the theme) for `publication`. With type set to `publication` in test-foo (located at content/foo/test-foo.md) it does not appear at mysiteurl/publication/test-foo/ (404) nor mysiteurl/publications/test-foo/ (404), and is not part of the list at mysiteurl/publications/ (renders a list with all content in the publications content directory).

> It would make sense that we would aim, as a theme development team, to develop in a way that pages reference the lowest point in the look-up order so that users can add increased specificity.

That seems correct. That would probably mean removing the `type = "post"` line from the default archetype and updating the example site? Can you think of anything else that would need to change to do that?

**Comment by pacollins** [comment]
I can't think of any other places, but I was also thinking out loud for the larger theme as a whole. If you could go ahead and do that, we can merge!

**Comment by matt9j** [comment]
> I can't think of any other places, but I was also thinking out loud for the larger theme as a whole. If you could go ahead and do that, we can merge!

Done!
